File: passport/src/store/vanityName.ts
/**
 * Vanity-name rules (doc 17), the namespace gate items 1-2. Pure, shared logic
 * for the future Findable mode: a name's charset, normalization, length, and the
 * reserved set. The server enforces the same rules at registration and the client
 * checks them before asking, so this is the single source of truth for both.
 *
 * Findable is not built yet; nothing calls this until the directory + resolve
 * endpoint land. It carries no status, no key, no identity, only the rules a name
 * must pass to be claimable.
 */

// Allowed name shape: lowercase letters, digits, underscore; length 3 to 30. No
// Unicode (removes homoglyph / confusable attacks at the namespace level).
const NAME_SHAPE = /^[a-z0-9_]{3,30}$/;
export const MIN_VANITY_LEN = 3;
export const MAX_VANITY_LEN = 30;

/**
 * Operational / official-impersonation terms that are unclaimable (doc 17). Brand
 * and service words, so a name cannot masquerade as sti.care or its staff. A
 * starter set, grown by report-and-takedown; kept lowercase to match normalized
 * input.
 */
export const RESERVED_NAMES: ReadonlySet<string> = new Set([
  "admin",
  "administrator",
  "root",
  "system",
  "sys",
  "official",
  "staff",
  "team",
  "mod",
  "moderator",
  "support",
  "help",
  "helpdesk",
  "contact",
  "info",
  "abuse",
  "security",
  "legal",
  "privacy",
  "billing",
  "payments",
  "api",
  "app",
  "www",
  "mail",
  "email",
  "noreply",
  "no_reply",
  "sti",
  "sticare",
  "care",
  "health",
  "clinic",
  "verify",
  "verified",
  "test",
  "null",
  "undefined",
]);

// The abuse / hate blocklist is NOT a client concern. The server holds the single
// authoritative, hate-only list (server/internal/vanityname) and answers a blocked
// name with a 409, so the client never ships its own divergent list. The client's
// instant check is charset + length + reserved only; the server is the source of
// truth for what is blocked. A client-side scam/phishing list (the old BLOCKED_NAMES)
// did not overlap the server's and risked pushing the app's own audience (identity /
// health / sexual terms) out of the namespace, so it was removed (doc 17).

// Normalize input to the canonical form names are stored and compared in: trim
// surrounding whitespace, lowercase. Shape validation still rejects anything that
// is not [a-z0-9_]; normalization only folds case and trims, it does not strip.
export function normalizeVanityName(input: string): string {
  return input.trim().toLowerCase();
}

/**
 * Why a (normalized) name fails the instant client check, or null when it passes.
 * There is no "blocked" reason: the hate blocklist is server-authoritative (a 409 at
 * registration), so the client only judges charset, length, and the reserved set.
 */
export type VanityNameError =
  "too-short" | "too-long" | "bad-chars" | "reserved";

/**
 * The instant client check: returns the first failing reason or null. Operates on the
 * normalized form, so callers should normalize first (this normalizes again,
 * defensively, so a raw value is also handled safely). It deliberately does NOT judge
 * whether a name is blocked for abuse; that is the server's call (the 409 path).
 */
export function vanityNameError(input: string): VanityNameError | null {
  const name = normalizeVanityName(input);
  if (name.length < MIN_VANITY_LEN) return "too-short";
  if (name.length > MAX_VANITY_LEN) return "too-long";
  if (!NAME_SHAPE.test(name)) return "bad-chars";
  if (RESERVED_NAMES.has(name)) return "reserved";
  return null;
}

/** Whether a name passes the instant client check (shape + length + not reserved). */
export function isValidVanityName(input: string): boolean {
  return vanityNameError(input) === null;
}

/**
 * Whether a value is a well-shaped vanity name (charset + length only), ignoring
 * the reserved/blocked sets. Used to validate a name PERSISTED in the account blob:
 * the charset is fixed, but the reserved/blocked sets grow over time, so validating
 * a stored name against them would brick an account that registered a name later
 * added to the blocklist (blob parse fails closed). Shape is the stable invariant.
 */
export function hasVanityNameShape(x: unknown): x is string {
  return typeof x === "string" && NAME_SHAPE.test(x);
}

File: server/internal/vanityname/vanityname.go
// Package vanityname holds the name rules for the Findable directory (doc 17):
// normalization, the legal charset/length, and the two unclaimable sets
// (reserved + blocklist). These are pure, deterministic checks with no I/O, so
// the registration endpoint and a periodic re-scan share one definition of
// "allowed" and a name can never be accepted one place and swept another.
//
// Findable is launched: the server wires these checks into the live registration,
// resolve, and report paths (GET/PUT/DELETE /u/{name} plus POST /u/{name}/report).
package vanityname

import (
	_ "embed"
	"errors"
	"regexp"
	"strings"
)

// The legal form after normalization: lowercase ASCII letters, digits, and
// underscore, 3 to 30 characters. No Unicode (doc 17): a name is exactly the
// bytes it looks like, which removes homoglyph/confusable attacks at the charset
// level. The resolve path normalizes again and re-checks defensively.
var formatRE = regexp.MustCompile(`^[a-z0-9_]{3,30}$`)

// Why a name was rejected. Returned by Check so the caller can surface a clear
// error at registration and so the re-scan can attribute an auto-takedown.
var (
	ErrFormat   = errors.New("vanityname: not lowercase [a-z0-9_]{3,30}")
	ErrReserved = errors.New("vanityname: reserved (operational/official term)")
	ErrBlocked  = errors.New("vanityname: blocklisted (abuse/impersonation)")
)

// Normalize maps raw input to its canonical form: trimmed and lowercased. It does
// NOT strip illegal characters; ValidFormat rejects anything left over, so a name
// is never silently rewritten into a different one.
func Normalize(raw string) string {
	return strings.ToLower(strings.TrimSpace(raw))
}

// ValidFormat reports whether an already-normalized name matches the legal
// charset and length. Callers pass the output of Normalize.
func ValidFormat(name string) bool {
	return formatRE.MatchString(name)
}

// Reserved operational / official-impersonation terms (doc 17 §63). A starter
// list, versioned here and grown by report-and-takedown; not exhaustive. Holding
// these unclaimable stops an account masquerading as the service or its staff.
var reservedList = []string{
	"admin", "administrator", "root", "system", "sys", "official", "staff",
	"team", "mod", "moderator", "support", "help", "helpdesk", "contact",
	"info", "abuse", "security", "legal", "privacy", "billing", "payments",
	"api", "app", "www", "mail", "email", "noreply", "no_reply", "sti",
	"sticare", "care", "health", "clinic", "verify", "verified", "test",
	"null", "undefined", "demo",
}

var reserved = toSet(reservedList)

// The abuse/impersonation blocklist (doc 17 §63): slurs and harassment terms,
// curated by the operator (and counsel) and grown reactively, kept in a data file
// (no .go edit needed) and deliberately out of the design doc. NOTE: the file is
// go:embed'd into the binary below, so a change to it only takes effect after a
// rebuild + redeploy; for an already-registered name that needs to come down
// immediately, the operator uses the runtime admin takedown endpoint (doc 20)
// instead. Authority look-alikes (`*_official`, `the_real_*`) are NOT auto-blocked
// here; those are handled reactively. The committed file may be empty until curated.
//
//go:embed blocklist.txt
var blocklistData string

var blocklist = parseTerms(blocklistData)

// Reserved reports whether a normalized name is in the reserved set.
func Reserved(name string) bool {
	_, ok := reserved[name]
	return ok
}

// Blocked reports whether a normalized name is in the blocklist.
func Blocked(name string) bool {
	_, ok := blocklist[name]
	return ok
}

// Check normalizes raw input and returns the canonical name plus the first rule
// it violates (nil err = allowed). The order is format, then reserved, then
// blocked, so the error is the most fundamental reason.
func Check(raw string) (name string, err error) {
	name = Normalize(raw)
	switch {
	case !ValidFormat(name):
		return name, ErrFormat
	case Reserved(name):
		return name, ErrReserved
	case Blocked(name):
		return name, ErrBlocked
	default:
		return name, nil
	}
}

// parseTerms reads a newline-separated term list: each line is normalized, blank
// lines and `#` comments are skipped. Defensive normalization means a file edited
// by hand can't smuggle in an uppercase or padded term that would never match.
func parseTerms(data string) map[string]struct{} {
	set := make(map[string]struct{})
	for _, raw := range strings.Split(data, "\n") {
		line := strings.TrimSpace(raw)
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}
		set[Normalize(line)] = struct{}{}
	}
	return set
}

func toSet(items []string) map[string]struct{} {
	set := make(map[string]struct{}, len(items))
	for _, it := range items {
		set[it] = struct{}{}
	}
	return set
}
